### PR TITLE
[Generator] Refactor extension methods and fix bug.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Moved the extension methods to their own file and enabled nullable.

Fixed the following underlying bug:

The extension method that creates a valid parameter named does not do the right thing in the following cases:

1. When the starting illegal char is NOT a number. It will prepend @ fixing nothing. Example " OHOH" to "@ OHOH"
2. When the illegal chars is in the middle of the param name. Example "OH OH" to "@OH OH"

I have fixed the method to return null in those ocassions (we will need to enable nullability later in the generator, is too much for this PR). Tests have been added to ensure we do not have such an issue again.

### CODE REVIEW

File: tests/generator/StringExtensionTests.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+				}
+			}
+		}
+		
+		[TestCaseSource(typeof(KeywordsDataSource))]
+		public void KeywordsTest(string keyword)
+			=> Assert.AreNotEqual (keyword, keyword.GetSafeParamName (), "keyword");
+
+		[TestCase ("😀OhOh")]
+		[TestCase (" OhOh")]
```
Comment (mandel-macaque) [inline tests/generator/StringExtensionTests.cs:37]: @haritha-mohan  has a point, "  OH_OH" would not per se be an error. But we are going to work with the fact that we will not get a space from bgen and use it as a test case.

**Comment by oh** [comment]
Hey

**Comment by mandel-macaque** [comment]
> Hey

LOL sorry that was a code example and the commit was done in the terminal, Github tried to be too smart.

File: src/generator-extension-methods.cs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+		if (!IsValidIdentifier (paramName, out var hasIllegalChars)) {
+			return hasIllegalChars ? null : "@" + paramName;
+		}
+		return paramName;
+	}
+
+	// Since we're building against the iOS assemblies and there's no code generation there,
+	// I'm bringing the implementation from:
+	// mono/mcs/class//System/Microsoft.CSharp/CSharpCodeGenerator.cs
+	static bool IsValidIdentifier (string? identifier, out bool hasIllegalChars)
```
Comment (haritha-mohan) [inline src/generator-extension-methods.cs:180]: I like the out keyword implementation- a cleaner way to keep track of what is going on!

**Review by haritha-mohan — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Review by rolfbjarne — APPROVED** [review]

**Comment by dalexsoto** [comment]
Generator diff is empty ✅ so we are happy!

**Pull request merged**